src/esp/io/JsonMagnumTypes.h

```diff
@@ -65,6 +65,68 @@ inline bool fromJsonValue(const JsonGenericValue& obj, Magnum::Vector3& val) {
   return false;
 }
 
+inline JsonGenericValue toJsonValue(const Magnum::Vector4& vec,
+                                    JsonAllocator& allocator) {
+  return toJsonArrayHelper(vec.data(), 4, allocator);
+}
+
+/**
+ * @brief Specialization to handle Magnum::Vector4 values. Populate passed @p
+ * val with value. Returns whether successfully populated, or not. Logs an error
+ * if inappropriate type.
+ *
+ * @param obj json value to parse
+ * @param val destination value to be populated
+ * @return whether successful or not
+ */
+inline bool fromJsonValue(const JsonGenericValue& obj, Magnum::Vector4& val) {
+  if (obj.IsArray() && obj.Size() == 4) {
+    for (rapidjson::SizeType i = 0; i < 4; ++i) {
+      if (obj[i].IsNumber()) {
+        val[i] = obj[i].GetDouble();
+      } else {
+        ESP_ERROR() << "Invalid numeric value specified in JSON Vec4, index :"
+                    << i;
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+inline JsonGenericValue toJsonValue(const Magnum::Color3& color,
+                                    JsonAllocator& allocator) {
+  return toJsonArrayHelper(color.data(), 3, allocator);
+}
+
+/**
+ * @brief Specialization to handle Magnum::Color3 values. Populate passed @p
+ * val with value. Returns whether successfully populated, or not. Logs an error
+ * if inappropriate type.
+ *
+ * @param obj json value to parse
+ * @param val destination value to be populated
+ * @return whether successful or not
+ */
+inline bool fromJsonValue(const JsonGenericValue& obj, Magnum::Color3& val) {
```

Comment:
To avoid code repetition you should be able to do something like this here, the `Color3`  is a direct subclass of `Vector3`:

```cpp
inline bool fromJsonValue(const JsonGenericValue& obj, Magnum::Color3& val) {
    return fromJsonValue(obj, static_cast<Magnum::Vector3&>(val));
}
```

Actually, it feels to me like this overload shouldn't be needed at all, the compiler could pick the `Vector3` variant on its own, I hope?

Same would be with the `Color4` variant below, which could now be taken care of with the `Vector4` variant, and I think the `toJsonValue()` for `Color3` and `Color4` wouldn't need to exist either?

(Personal feeling and [not a shameless self-promo at all](https://doc.magnum.graphics/corrade/classCorrade_1_1Utility_1_1Json.html#Utility-Json-usage-direct-array-access): wow, rapidjson is *verbose*...)

Reply:
> Actually, it feels to me like this overload shouldn't be needed at all, the compiler could pick the Vector3 variant on its own, I hope?

The compiler picks `Color3` 🙁.

> (Personal feeling and [not a shameless self-promo at all](https://doc.magnum.graphics/corrade/classCorrade_1_1Utility_1_1Json.html#Utility-Json-usage-direct-array-access): wow, rapidjson is verbose...)

We definitely could look into using Corrade for this. It might ease up future extensibility, especially as we expect to expand these functionalities. It would also take out one dependency.

Reply:
> The compiler picks Color3 :slightly_frowning_face:.

Ah well, so if you don't have the `fromJsonValue(..., Color3&)` overload at all, it'll fail to compile? Strange, I thought it would just go with the base class. Or maybe my understanding of how overload resolution works for mutable references is wrong.

In any case, the explicit `static_cast` would reduce the code duplication at least partially.